@let bug fixes

crisbeto opened the pull request:
Includes the following fixes related to `@let`:

### fix(compiler-cli): type check let declarations nested inside nodes
Fixes that we were only capturing `@let` declarations at the top level of the scope, not any of the nested children.

### fix(compiler-cli): flag all conflicts between let declarations and local symbols
Expands the check around conflicting `@let` declarations to also cover template variables and local references.

### fix(compiler): give precedence to local let declarations over parent ones
Currently the logic that maps a name to a variable looks at the variables in their definition order. This means that `@let` declarations from parent views will always come before local ones, because the local ones are declared inline whereas the parent ones are hoisted to the top of the function.

These changes resolve the issue by giving precedence to the local variables.

Fixes https://github.com/angular/angular/issues/56737.

crisbeto committed:
fix(compiler-cli): type check let declarations nested inside nodes

Fixes that we were only capturing `@let` declarations at the top level of the scope, not any of the nested children.

crisbeto committed:
fix(compiler-cli): flag all conflicts between let declarations and local symbols

Expands the check around conflicting `@let` declarations to also cover template variables and local references.

crisbeto commented on packages/compiler/src/template/pipeline/src/phases/resolve_names.ts:
```diff
@@ -36,6 +36,9 @@ function processLexicalScope(
   const scope = new Map<string, ir.XrefId>();
 
+  // Symbols defined within the current scope. They take precedence over ones defined outside.
+  const localDefinitions = new Map<string, ir.XrefId>();
```
I don't think that this is the most elegant way of solving it, but I think it's likely the most future-proof. Some alternatives that were considered, but didn't work:
1. Iterating the instructions in reverse seems to break cases where variables depend on other variables. Otherwise this would've been the simplest.
2. We can't hoist the local `@let` variables to the top, because they need an `advance` call.

thePunderWoman reviewed (approved):
reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

alxhub commented on packages/compiler/src/template/pipeline/src/phases/resolve_names.ts:
```diff
@@ -46,10 +49,14 @@ function processLexicalScope(
-              continue;
+            if (
+              !localDefinitions.has(op.variable.identifier) &&
+              op.initializer instanceof ir.StoreLetExpr
```
Ideally the resolve_names phase wouldn't have any special knowledge of `@let` semantics. I think you're right that we do need the concept of "local declaration" though, which shadows any variables of the same name that might've been discovered before it.

I suggest we encode this into the `IdentifierVariable` structure. Have a `local` property which is set to `true` when generate_local_let_references adds the variable and `false` for other `IdentifierVariable`s.

crisbeto committed:
fix(compiler): give precedence to local let declarations over parent ones

Currently the logic that maps a name to a variable looks at the variables in their definition order. This means that `@let` declarations from parent views will always come before local ones, because the local ones are declared inline whereas the parent ones are hoisted to the top of the function.

These changes resolve the issue by giving precedence to the local variables.

Fixes #56737.

thePunderWoman commented:
This PR was merged into the repository by commit 2a1291e942a3cd645ee635e72e7d83722383d39b.

The changes were merged into the following branches: main
